File: app/src/main/java/com/sq26/experience/ui/activity/file/FileOperateActivity.java
package com.sq26.experience.ui.activity.file;

import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.text.format.Formatter;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.documentfile.provider.DocumentFile;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.sq26.experience.R;
import com.sq26.experience.adapter.RecyclerViewListAdapter;
import com.sq26.experience.adapter.ViewHolder;
import com.sq26.experience.util.FileUtil;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FileOperateActivity extends AppCompatActivity {
    //父文件夹列表视图
    @BindView(R.id.rootRecyclerView)
    RecyclerView parentRecyclerView;
    //文件列表视图
    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;

    //父文件夹列表适配器
    private RecyclerViewListAdapter<DocumentFile> parentFileListAdapter;
    //父文件夹列表
    private List<DocumentFile> parentFileList = new ArrayList<>();
    //文件列表适配器
    private RecyclerViewListAdapter<DocumentFile> fileListAdapter;
    //文件列表
    private List<DocumentFile> fileList = new ArrayList<>();
    //上下文
    private Context context;
    //其他储存的写权限申请码
    private final int DOCUMENT_TREE_CODE = 0;
    //主储存的根目录路径
    private String rootPath = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_file_management);
        ButterKnife.bind(this);
        context = this;
        //获取根目录路径
        rootPath = getIntent().getStringExtra("rootPath");
        //初始化视图
        initView();

        Log.d("DIRECTORY_MUSIC", getExternalFilesDir(Environment.DIRECTORY_MUSIC).getAbsolutePath());
        Log.d("DIRECTORY_PODCASTS", getExternalFilesDir(Environment.DIRECTORY_PODCASTS).getAbsolutePath());
        Log.d("DIRECTORY_RINGTONES", getExternalFilesDir(Environment.DIRECTORY_RINGTONES).getAbsolutePath());
        Log.d("DIRECTORY_ALARMS", getExternalFilesDir(Environment.DIRECTORY_ALARMS).getAbsolutePath());
        Log.d("DIRECTORY_NOTIFICATIONS", getExternalFilesDir(Environment.DIRECTORY_NOTIFICATIONS).getAbsolutePath());
        Log.d("DIRECTORY_PICTURES", getExternalFilesDir(Environment.DIRECTORY_PICTURES).getAbsolutePath());
        Log.d("DIRECTORY_MOVIES", getExternalFilesDir(Environment.DIRECTORY_MOVIES).getAbsolutePath());
        Log.d("NULL", getExternalFilesDir(null).getAbsolutePath());
    }

    private void initView() {
        parentFileListAdapter = new RecyclerViewListAdapter<DocumentFile>(R.layout.item_parent_file, parentFileList) {
            @Override
            public void bindViewHolder(ViewHolder viewHolder, DocumentFile item, int position) {
                viewHolder.setText(R.id.text, item.getName());
                viewHolder.setOnClickListener(R.id.text, view -> {
                    parentFileList.subList(position + 1, parentFileList.size()).clear();
                    notifyDataSetChanged();
                    initData(item.listFiles());
                });
            }
        };
        parentRecyclerView.setAdapter(parentFileListAdapter);

        //文件列表适配器
        fileListAdapter = new RecyclerViewListAdapter<DocumentFile>(R.layout.item_file, fileList) {
            @Override
            public void bindViewHolder(ViewHolder viewHolder, DocumentFile item, int position) {
                viewHolder.setText(R.id.name, item.getName());
                viewHolder.setTextColor(R.id.name, item.canWrite() ? Color.GREEN : Color.RED);
                viewHolder.setText(R.id.dateTime, new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(item.lastModified()));
                SimpleDraweeView simpleDraweeView = viewHolder.getView(R.id.simpleDraweeView);
                if (item.isFile()) {
                    simpleDraweeView.setImageResource(R.drawable.ic_insert_drive_file_black_24dp);
                    viewHolder.setText(R.id.remark, Formatter.formatFileSize(context, item.length()));
                } else {
                    simpleDraweeView.setImageResource(R.drawable.ic_folder_open_black_24dp);
                    viewHolder.setText(R.id.remark, item.listFiles().length + "个项目");
                }
                viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (item.isFile()) {
                            FileUtil.openFile(context, item.getUri());
                        } else {
                            parentFileList.add(item);
                            parentFileListAdapter.notifyDataSetChanged();
                            initData(item.listFiles());
                        }
                    }
                });
                viewHolder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View view) {
                        new AlertDialog.Builder(FileOperateActivity.this)
                                .setMessage("是否删除")
                                .setPositiveButton("是", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialogInterface, int i) {
                                        Log.d("删除", "有权限");
                                        item.delete();
                                        fileList.remove(item);
                                        fileListAdapter.notifyDataSetChanged();
                                    }
                                }).show();
                        return true;
                    }
                });
            }
        };
        //设置分割线
        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        //设置适配器
        recyclerView.setAdapter(fileListAdapter);
        //声明一个DocumentFile对象
        DocumentFile documentFile;
        //判断是否是绝对路径
        if (FileUtil.isAbsolutePath(rootPath)) {
            //是绝对路径,要先转成File类
            documentFile = DocumentFile.fromFile(new File(rootPath));
        } else {
            //不是绝对路径
            documentFile = DocumentFile.fromTreeUri(context, Uri.parse(rootPath));
        }
        parentFileList.add(documentFile);
        parentFileListAdapter.notifyDataSetChanged();
        //初始化数据
        initData(documentFile.listFiles());
    }

    private void initData(DocumentFile[] documentFiles) {
        //先清理空文件列表
        fileList.clear();
        //直接将DocumentFile数值转化成DocumentFile列表,全部添加进文件列表
        fileList.addAll(Arrays.asList(documentFiles));
        //刷新文件列表视图
        fileListAdapter.notifyDataSetChanged();
    }

    @Override
    public void onBackPressed() {
        if (parentFileList.size() > 1) {
            parentFileList.remove(parentFileList.size() - 1);
            parentFileListAdapter.notifyDataSetChanged();
            initData(parentFileList.get(parentFileList.size() - 1).listFiles());
        } else {
            super.onBackPressed();
        }
    }

}

File: app/src/main/java/com/sq26/experience/ui/activity/file/FileOperateActivity.kt
package com.sq26.experience.ui.activity.file

import android.content.ContentResolver
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.activity.addCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.net.toFile
import androidx.core.net.toUri
import androidx.databinding.DataBindingUtil
import androidx.documentfile.provider.DocumentFile
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DividerItemDecoration
import com.sq26.experience.R
import com.sq26.experience.adapter.CommonListAdapter
import com.sq26.experience.adapter.CommonListViewHolder
import com.sq26.experience.databinding.ActivityFileManagementBinding
import com.sq26.experience.databinding.ItemFileBinding
import com.sq26.experience.databinding.ItemParentFileBinding
import com.sq26.experience.util.FileUtil
import java.util.*

class FileOperateActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        DataBindingUtil.setContentView<ActivityFileManagementBinding>(
            this,
            R.layout.activity_file_management
        )
            .apply {
                lifecycleOwner = this@FileOperateActivity

                //父文件夹列表
                val parentFileList: MutableList<DocumentFile> = ArrayList()
                //创建父文件夹目录适配器
                val parentFileListAdapter =
                    object : CommonListAdapter<DocumentFile>(DocumentFileDiffCallback()) {
                        override fun createView(
                            parent: ViewGroup,
                            viewType: Int
                        ): CommonListViewHolder<*> {
                            //创建通用视图持有者
                            return object : CommonListViewHolder<ItemParentFileBinding>(
                                ItemParentFileBinding.inflate(
                                    LayoutInflater.from(parent.context),
                                    parent,
                                    false
                                )
                            ) {
                                //绑定数据
                                override fun bind(position: Int) {
                                    //设置参数
                                    v.documentFile = getItem(position)
                                    //设置点击事件
                                    v.setOnClick {
                                        //移除从选定位置到结尾的目录
                                        parentFileList.subList(position + 1, parentFileList.size)
                                            .clear()
                                        //刷新视图
                                        submitList(parentFileList.toList())
                                    }
                                    //强制绑定数据
                                    executePendingBindings()
                                }
                            }
                        }

                    }
                //绑定适配器
                rootRecyclerView.adapter = parentFileListAdapter

                //文件列表
                val fileList: MutableList<DocumentFile> = ArrayList()
                //创建文件适配器
                val fileListAdapter =
                    object : CommonListAdapter<DocumentFile>(DocumentFileDiffCallback()) {
                        override fun createView(
                            parent: ViewGroup,
                            viewType: Int
                        ): CommonListViewHolder<*> {
                            return object : CommonListViewHolder<ItemFileBinding>(
                                ItemFileBinding.inflate(
                                    LayoutInflater.from(parent.context),
                                    parent,
                                    false
                                )
                            ) {
                                init {
                                    //设置根目录的点击事件
                                    v.root.setOnClickListener {
                                        v.item?.let {
                                            if (it.isFile) {
                                                FileUtil.openFile(
                                                    this@FileOperateActivity,
                                                    it.uri
                                                )
                                            } else {
                                                parentFileList.add(it)
                                                parentFileListAdapter.submitList(parentFileList.toList())
                                                submitList(it.listFiles().toList())
                                            }
                                        }
                                    }
                                    //设置根目录的长按事件
                                    v.root.setOnLongClickListener {
                                        v.item?.let {
                                            AlertDialog.Builder(this@FileOperateActivity)
                                                .setMessage("是否删除")
                                                .setPositiveButton("是") { _, _ ->
                                                    Log.d("删除", "有权限")
                                                    //删除数据
                                                    it.delete()
                                                    //移除对应的documentFile对象
                                                    fileList.remove(it)
                                                    //刷新视图
                                                    submitList(fileList.toList())
                                                }.show()
                                        }
                                        true
                                    }
                                }
                                override fun bind(position: Int) {
                                    v.item = getItem(position)
                                }
                            }
                        }

                    }
                //设置分割线
                recyclerView.addItemDecoration(
                    DividerItemDecoration(
                        this@FileOperateActivity,
                        DividerItemDecoration.VERTICAL
                    )
                )
                //绑定适配器
                recyclerView.adapter = fileListAdapter
                //获取根目录路径
                val rootUri = intent.getStringExtra("rootUri").orEmpty()
                //声明一个DocumentFile对象
                val documentFile =if (rootUri.toUri().scheme == ContentResolver.SCHEME_FILE){
                    DocumentFile.fromFile(rootUri.toUri().toFile())
                }else{
                    DocumentFile.fromTreeUri(this@FileOperateActivity, Uri.parse(rootUri))
                }

                documentFile?.let {
                    //加入父目录列表
                    parentFileList.add(it)
                    //刷新父目录视图
                    parentFileListAdapter.submitList(parentFileList.toList())
                    //初始化数据
                    fileListAdapter.submitList(it.listFiles().toList())
                }
                //添加返回键监听
                onBackPressedDispatcher.addCallback {
                    //parentFileList数量大于1说明不是根目录
                    if (parentFileList.size > 1) {
                        //移除最后一个目录
                        parentFileList.removeAt(parentFileList.size - 1)
                        //刷新父目录列表
                        parentFileListAdapter.submitList(parentFileList.toList())
                        //刷新文件列表
                        fileListAdapter.submitList(listOf(*parentFileList[parentFileList.size - 1].listFiles()))
                    } else {
                        //关闭返回监听
                        isEnabled = false
                        //主动调用返回
                        onBackPressedDispatcher.onBackPressed()
                        onBackPressedDispatcher.hasEnabledCallbacks()
                    }
                }
            }
    }
}

class DocumentFileDiffCallback : DiffUtil.ItemCallback<DocumentFile>() {
    //判断唯一标识是否相同
    override fun areItemsTheSame(
        oldItem: DocumentFile,
        newItem: DocumentFile
    ): Boolean {
        return oldItem.uri == newItem.uri
    }

    //判断内容是否相同
    override fun areContentsTheSame(
        oldItem: DocumentFile,
        newItem: DocumentFile
    ): Boolean {
        return oldItem.uri == oldItem.uri
    }

}